ui/js/components/auth.js

```diff
@@ -0,0 +1,165 @@
+"use strict";
+
+/**
+ * This component handles logging in to Taskcluster Authentication
+ *
+ * See: https://docs.taskcluster.net/manual/3rdparty
+ *
+ * This communicates to the rest of Treeherder by calling the onUserChange
+ * function on the HTML element, which in turn stores that user in $rootScope.
+ */
+treeherder.component("login", {
+    template: `
+        <span class="dropdown"
+              ng-if="$ctrl.user.loggedin">
+          <button id="logoutLabel" title="Logged in as: {{$ctrl.user.email}}" role="button"
+                  data-toggle="dropdown" data-target="#"
+                  class="btn btn-view-nav btn-right-navbar nav-persona-btn">
+            <div class="nav-user-icon">
+              <span class="fa fa-user pull-left"></span>
+            </div>
+            <span class="fa fa-angle-down lightgray"></span>
+          </button>
+          <ul class="dropdown-menu" role="menu" aria-labelledby="logoutLabel">
+              <li>
+                <a ng-click="$ctrl.logout()">Logout</a>
+              </li>
+          </ul>
+        </span>
+
+        <a class="btn btn-view-nav btn-right-navbar nav-login-btn"
+           ng-if="!$ctrl.user.loggedin"
+           ng-click="$ctrl.login()">Login/Register</a>
+    `,
+    bindings: {
+        // calls to the HTML which sets the user value in the $rootScope.
+        onUserChange: "&"
+    },
+    controller: ['$location', '$window', 'localStorageService',
+        'ThUserModel', '$http', 'thUrl', '$timeout',
+        function ($location, $window, localStorageService,
+                  ThUserModel, $http, thUrl, $timeout) {
+            var ctrl = this;
+            ctrl.user = {};
+            // "clears out" the user when it is detected to be logged out.
+            var loggedOutUser = {is_staff: false, username: "", email: "", loggedin: false};
+
+            /**
+             * Using a window listener here because I wasn't getting reliable
+             * results from the events from angular-local-storage.
+             */
+            $window.addEventListener("storage", function(e) {
```

Comment:
Ah cool - so multiple tabs will all pick up a login without needing a reload - great! (Does that even work at the moment with Persona?)